lc0/src/neural/network_blas_cl.h

```diff
@@ -0,0 +1,176 @@
+members are those used exclusively by the CPU implementation.
+*/
+
+#pragma once
+
+#include "factory.h" // network.h, optionsdict.h
+#include "blas_config.h"
+#include "opencl/OpenCLScheduler.h"
+
+namespace lczero {
+
+class BlasNetwork;
+
+class BlasNetworkComputation : public NetworkComputation {
+
+ public:
+  BlasNetworkComputation(BlasNetwork* network) : network(network) {}
+
+  void AddInput(InputPlanes&& input) override {
+    inputs.emplace_back(input);
+  }
+
+  int GetBatchSize() const override {
+    return inputs.size();
+  }
+
+  float GetQVal(int sample) const override {
+    return output_values[sample];
+  }
+
+  float GetPVal(int sample, int move_id) const override {
+    return output_policies[sample][move_id];
+  }
+
+  void ComputeBlocking() override;
+
+ private:
+  std::vector<InputPlanes> inputs;
+  std::vector<float>   output_values;
+  std::vector<std::vector<float>> output_policies;
+  BlasNetwork* network;
+};
+
+class BlasNetwork : public Network {
+
+ public:
+  BlasNetwork(const Weights& weights, const OptionsDict& options);
+
+  std::unique_ptr<NetworkComputation> NewComputation() override {
+    return std::make_unique<BlasNetworkComputation>(this);
+  }
+
+  std::pair<float, std::vector<float>> evaluate(InputPlanes& input);
+  // this is the main function used by outside code
+
+ protected: // shared functions between blas and OpenCL implementations
+  virtual void forwardPass(const std::vector<float>& input,
+                                 std::vector<float>& policy_data,
+                                 std::vector<float>& value_data);
+  // forwardPass is the actual network computation; evaluate() wraps this.
+  static std::vector<float> innerproduct(const std::vector<float>& inputs,
+                                         const std::vector<float>& weights,
+                                         const std::vector<float>& biases,
+                                         bool apply_relu=false);
+  /* TODO: understand why gcp left the final value head computation out of OpenCL,
+  leaving it to be done on the cpu. This is why forwardPass() doesn't simply return
+  the evaluation, rather than the value_data array which needs to be processed by
+  innerproduct(). If this last layer were done in OpenCL, then innerproduct() would
+  not need to be shared. Best guess so far: for selfcheck purposes, it's easier
+  to debug/compare the value head's second-to-last output vector, rather than
+  its last output scalar?
+  */
+
+  void initOneBlock(Weights::ConvBlock& block, bool inputlayer=false);
+  static std::vector<float> winograd_transform_f(const std::vector<float>& f, const int outputs, const int channels);
+  static std::vector<float> softmax(const std::vector<float>& input, float temperature=1.0f);
+  // TODO: softmaxtemp hardcoded from lczero
+
+  Weights weights_; // optimal memory use? is one reference shared among multiple backends?
```

Comment:
Is it modified by any of computations? If no, `const`?

Reply:
yes it is, initOneBlock(). If it was const, it could merely be a reference instead of its own copy